### PR TITLE
Adding support for allowed groups in SAML Connector

### PR DESCRIPTION
**Pull request by kenperkins**
I'm not suggesting this should be taken as is, but I wanted to use it as a conversation starter on adding SAML support for group filtering. This is based on conversation in #1476 

I thought briefly about a generalized approach but right now I just wanted to get a primitive implementation vetted and ideally merged, and then follow that with figuring out what it looks like across providers.

Obviously, there are no tests or docs in this change.

### CODE REVIEW

**Comment by kenperkins** [comment]
I've now added a few tests that test the groups behavior for SAML:

- TestGroupsWhitelist (User should be allowed)
- TestGroupsWhitelistEmpty (Empty groups whitelist)
- TestGroupsWhitelistDisallowed (User not allowed)
- TestGroupsWhitelistDisallowedNoGroupsOnIdent (User has no groups from saml backend)

**Comment by kenperkins** [comment]
I've added docs with my latest commit.

File: connector/saml/saml.go [hunk 1]
```diff
@@ -408,6 +415,30 @@ func (p *provider) HandlePOST(s connector.Scopes, samlResponse, inResponseTo str
...
 		ident.Groups = groups
 	}
+
+	if len(p.allowedGroups) == 0 {
+		// No allowed groups set, just return the ident
+		return ident, nil
+	}
+
+	// Look for membership in one of the allowed groups
+	allowed := false
```
Comment (srenatus) [inline connector/saml/saml.go]: Can we use the `groups.Filter` utility method here?

Comment (kenperkins) [reply connector/saml/saml.go]: I hadn't seen that method, will investigate.

Comment (kenperkins) [reply connector/saml/saml.go]: I've updated to use `groups.Filter`.

**Comment by srenatus** [comment]
A bird in the hand is worth two in the bush, so let's get this in and keep the refactor briefly discussed in #1476 in mind.

@kenperkins I'd like to get another maintainer to approve this; and would you mind squashing these commits? 😃

**Comment by kenperkins** [comment]
You need me to squash before merge? Do you not squash/merge as part of GH merge button? (That's fine, just looking for clarity).

**Comment by srenatus** [comment]
Usually just hitting merge 😉

**Comment by kenperkins** [comment]
I'm not sure you're waiting on me, I'm totally ok with my commits to be squashed, if you need me to do it, please give me clear guidance and I'll re-push.

**Comment by srenatus** [comment]
@kenperkins sorry, was on thumbs. I'd appreciate if **you could squash them**. Thanks! There's a bit of commit guidance in the dev docs: https://github.com/dexidp/dex/blob/60264d440cd668fe601f8ade81e100c8cd95e7f1/Documentation/dev-dependencies.md#composing-commits -- but we haven't been too strict in following these lately 😉 

Thanks again for your quick action here 🚀

**Comment by kenperkins** [comment]
So updated.

**Comment by kenperkins** [comment]
Is there anything I can do here to help?

**Comment by kenperkins** [comment]
@srenatus is there another maintainer that could assist in a review here?

**Comment by srenatus** [comment]
@kenperkins I suppose so! 🏓 @JoelSpeed @bonifaido could you please have a look?

File: connector/saml/saml_test.go [hunk 2]
```diff
@@ -98,6 +99,96 @@ func TestGroups(t *testing.T) {
 	test.run(t)
 }
 
+func TestGroupsWhitelist(t *testing.T) {
```
Comment (bonifaido) [inline connector/saml/saml_test.go]: Love the tests! 🎉

**Comment by kenperkins** [comment]
:tada:

**Comment by Sanyaorg** [comment]
Hey @kenperkins, do you have working config for allowedGroups? 
It doesn't work for me, so I'm not sure what the issue is. 
It still returns token with all groups inside.
Thanks
Appreciate your help

**Pull request merged**